### PR TITLE
Unknown tests

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This adds GRPC around unknowns for both attributes and collections for both PlanResourceChange and non-PlanResourceChange

### CODE REVIEW

**Review by iwahbe — COMMENTED** [review]
Thanks for building this out!

To clarify, do these tests assert on that the current behavior doesn't change, or that the current behavior matches TF's behavior?

I know it's a pain, but it would be really nice to have these tests as a table driven test:

```go
tests := []struct{
	input    resource.PropertyValue
	expected *resource.PropertyValue
}{
	{
		input: resource.NewComputed(resource.Property("")),
		expected: nil,
	},
	...
}
```

Looking through these tests, it looks like only the property value for input and output has changed, but it's hard to verify that for all of them. The rest makes it harder to understand what's going on. I have no idea if there is a relationship between the "normal" and PlanResource tests. It looks like they are the same tests, but with different responses.

By folding PF and "normal" testing into the same table, we would be able to see clearly where PlanResource diverges from `expected`, which I can't really tell without manually comparing tests now.

```patch
 tests := []struct{
 	input                resource.PropertyValue
 	expected             *resource.PropertyValue
+	expectedPlanResource *resource.PropertyValue
 }{

**Comment by VenelinMartinov** [comment]
Both of these assert on the current behaviour for both plan resource change and non-plan resource change.

I built it out to compare behaviour in https://github.com/pulumi/pulumi-terraform-bridge/pull/1971 and make sure we aren't regressing.

I'd like to merge as is and leave refactoring for later as https://github.com/pulumi/pulumi-terraform-bridge/pull/1971 depends on this and it has been hanging for quite a while

**Review by iwahbe — APPROVED** [review]
> Both of these assert on the current behaviour for both plan resource change and non-plan resource change.
> 
> I built it out to compare behaviour in #1971 and make sure we aren't regressing.

Can you add a comment explaining that these are regression tests but that we haven't verified that the behavior matches TF for each case then.

> I'd like to merge as is and leave refactoring for later as #1971 depends on this and it has been hanging for quite a while

That makes sense.

> The structure of the tests it that for each of the property it injects an unknown at every possible level and asserts on the response.

A comment explaining that would be great.

**Comment by VenelinMartinov** [comment]
GH token rate limits ❤️

File: pkg/tfbridge/provider_test.go [hunk 1]
```diff
@@ -4535,3 +4535,1072 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
...
+			"properties":{
+				"__defaults":[],
+				"nestedListProps":[["04da6b54-80e4-46f7-96ec-b56ff0331ba9"]]
+			},
+			"preview":true
+		},
+		"response": {
+			"properties":{
+				"id":"",
+				"nestedListProps":["04da6b54-80e4-46f7-96ec-b56ff0331ba9"]
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:4865]: Please comment on every unknown floating case that per our investigation this happens inside PlanResourceChange call and must be an artifact of how TF processes things internally.

File: pkg/tfbridge/provider_test.go [hunk 2]
```diff
@@ -4535,3 +4535,1072 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
...
+
+	t.Run("unknown for nested list block prop nested collection", func(t *testing.T) {
+		testutils.Replay(t, provider, `
+	{
+		"method": "/pulumirpc.ResourceProvider/Create",
+		"request": {
+			"urn": "urn:pulumi:dev::teststack::ExampleResource::exres",
+			"properties":{
+				"__defaults":[],
+				"nestedListBlockProps":[{"nestedProps":"04da6b54-80e4-46f7-96ec-b56ff0331ba9"}]
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:4921]: This special case deserves a comment that an unknown is being transformed to a null in an attempt to avoid sending unknowns where TF does not support them. Still this probably is a bug as far as the provider's concerned. 

Without a comment it looks like it's asserting behavior we want.

File: pkg/tfbridge/provider_test.go [hunk 3]
```diff
@@ -4535,3 +4535,1072 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
...
+		},
+		"response": {
+			"properties":{
+				"id":"04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+				"stringProp":null,
+				"setProps":null,
+				"listProps":null,
+				"nestedListProps":null,
+				"maxItemsOneProp":null,
+				"setBlockProps":[{"prop":""}],
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:5180]: This behavior is probably in error..

File: pkg/tfbridge/provider_test.go [hunk 4]
```diff
@@ -4535,3 +4535,1072 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
...
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:5209]: https://github.com/pulumi/pulumi-terraform-bridge/issues/1885

File: pkg/tfbridge/provider_test.go [hunk 5]
```diff
@@ -4535,3 +4535,1072 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
...
+		"response": {
+			"properties":{
+				"id":"04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+				"stringProp":null,
+				"setProps":null,
+				"listProps":null,
+				"nestedListProps":null,
+				"maxItemsOneProp":null,
+				"setBlockProps":[],
+				"listBlockProps":[{ "prop": null }],
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:5329]: https://github.com/pulumi/pulumi-terraform-bridge/issues/1885

**Comment by t0yv0** [comment]
@iwahbe we need some plumbing work to do to allow cross-tests to express unknowns in the meanwhile we test at this level. I requested a few additional comments/cross-links but it looks good. It was very important to me to have some edge cases covered here.

**Review by t0yv0 — APPROVED** [review]

**Comment by VenelinMartinov** [comment]
I've done some further digging around the results here in https://github.com/pulumi/pulumi-terraform-bridge/issues/2032 and https://github.com/pulumi/pulumi-terraform-bridge/pull/2060/files and https://github.com/pulumi/pulumi-terraform-bridge/pull/2060/files

I suspect we can do better with unknowns
